### PR TITLE
Added bwa-mem and index rule. Added testing fastq file and changed te…

### PR DESCRIPTION
**Pull request by jonca79**
…sting config. Added files to rule all.

### CODE REVIEW

File: workflow/rules/bwa-mem.smk [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+__author__ = "Jonas Almlöf"
+__copyright__ = "Copyright 2021, Jonas Almlöf"
+__email__ = "jonas.almlof@scilifelab.uu.se"
+__license__ = "GPL-3"
+
+
+rule bwa_mem:
+    input:
+        fastq=["Pre-alignment/merge/{sample}_R1.fq.gz", "Pre-alignment/merge/{sample}_R2.fq.gz"],
```
Comment (marrip) [inline workflow/rules/bwa-mem.smk:12]: I would be set fwd and rev as different stanza to make it more robust,
these should also be put in two lines I think according to snakefmt? It makes reading your code much easier

File: workflow/rules/bwa-mem.smk [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    log:
+        log="Logs/Aligment/bwa-mem/{sample}.log",
+    params:
+        index=config["references"]["genome"],
+        extra=config["bwa-mem"]["extra"],
+        sort=config["bwa-mem"].get("sort", "samtools")
+        sort_order=config["bwa-mem"].get("sort_order", "coordinate"),
+        sort_extra="-@ " + str(config["bwa-mem"]["threads"]),
+    threads: config["bwa-mem"]["threads"]
+    #benchmark:
```
Comment (marrip) [inline workflow/rules/bwa-mem.smk:24]: I am personally not a fan of commented code. But you probably use it at a later stage, right?

File: workflow/rules/bwa-mem.smk [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+rule bwa_mem:
+    input:
+        fastq=["Pre-alignment/merge/{sample}_R1.fq.gz", "Pre-alignment/merge/{sample}_R2.fq.gz"],
+    output:
+        bam="Alignment/bwa-mem/{sample}.bam",
+    log:
+        log="Logs/Aligment/bwa-mem/{sample}.log",
+    params:
+        index=config["references"]["genome"],
```
Comment (marrip) [inline workflow/rules/bwa-mem.smk:18]: Files should be part of input to have snakemake check that they are indeed available

File: workflow/rules/bwa-mem.smk [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        "{rule}: Align {wildcards.sample} with bwa and sort"
+    wrapper:
+        "0.70.0/bio/bwa/mem"
+
+
+rule index_bwa_bam:
+    input:
+        "Alignment/bwa-mem/{sample}.bam"
+    output:
+        "Alignment/bwa-mem/{sample}.bam.bai",
```
Comment (marrip) [inline workflow/rules/bwa-mem.smk:38]: I think the output should be in `index_bwa_bam` to have a link to the rule name or the log should be in `bwa-mem` but I guess there are certain problems with the naming conventions. I would also prefer if we use underscores everywhere and not a mixture to reduce confusion.

File: workflow/rules/bwa-mem.smk [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    #    repeat(_bwa_benchmark, config.get("benchmark", {}).get("repeats", 1))
+    container:
+        config["tools"].get("bwa-mem", config["tools"].get("default", ""))
+    message:
+        "{rule}: Align {wildcards.sample} with bwa and sort"
+    wrapper:
+        "0.70.0/bio/bwa/mem"
+
+
+rule index_bwa_bam:
```
Comment (marrip) [inline workflow/rules/bwa-mem.smk:34]: Maybe this rule name can be more general such as `samtools_index` and then reusable whenever we need to get an index for a bam file?

File: workflow/rules/common.smk [hunk 6]
```diff
@@ -50,5 +50,9 @@ def get_sample_fastq(wildcards):
     fastqs = units.loc[(wildcards.sample, wildcards.unit), ["fastq1", "fastq2"]].dropna()
     return {"fwd": fastqs["fastq1"].tolist(), "rev": fastqs["fastq2"].tolist()}
 
+
 def compile_output_list(wildcards):
-    return ["dummy.txt"]
+    output_list = []
+    output_list += expand("/Alignment/bwa-mem/{sample}.bam", sample=samples["sample"])
+    output_list += expand("/Alignment/bwa-mem/{sample}.bam.bai", sample=samples["sample"])
```
Comment (marrip) [inline workflow/rules/common.smk:57]: I always used `samples.index()`  to do the same thing but I am not sure if there are any advantages/disadvantages for one or the other

**Review by marrip — COMMENTED** [review]
We need to include unit as well in the file names/paths to have access to it downstream when we need it.

File: workflow/rules/bwa-mem.smk (hunk 1 above)
Comment (jonca79) [reply workflow/rules/bwa-mem.smk:12]: The wrapper needs it this way. Otherwise we will have to skip the wrapper.
https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/bwa/mem.html

File: workflow/rules/bwa-mem.smk (hunk 3 above)
Comment (jonca79) [reply workflow/rules/bwa-mem.smk:18]: Also how it is specified in the wrapper

File: workflow/rules/bwa-mem.smk (hunk 5 above)
Comment (jonca79) [reply workflow/rules/bwa-mem.smk:34]: Should this rule be in a separate module (Misc) or in the tools module or somewhere else?

File: workflow/rules/common.smk (hunk 6 above)
Comment (jonca79) [reply workflow/rules/common.smk:57]: I think Patrik also uses samples.index()  so lets choose this then

File: workflow/rules/bwa-mem.smk (hunk 4 above)
Comment (jonca79) [reply workflow/rules/bwa-mem.smk:38]: Changed the bwa-mem to just bwa. 
In general I agree but in this case I don't. I guess if we make the indexing function more general this will solve the issue.

Comment (marrip) [reply workflow/rules/bwa-mem.smk:38]: I guess it would. Still we might need to find some standard of naming things. I am also concerned about the misc rules and the separate log directory. If we make them very unspecific such as:

```
input:
  {file}.bam
output:
  {file}.bai
```

The separate log directory might not work as the the `log` stanza needs to have the same wildcards I believe. It would be easier to do:

```
log:
  {file}_index.log
```

File: workflow/rules/bwa-mem.smk (hunk 5 above)
Comment (marrip) [reply workflow/rules/bwa-mem.smk:34]: I think `Misc` would be a good choice. Tools seems to be a toolbox for building modules, right @Smeds ?

File: workflow/rules/bwa-mem.smk (hunk 2 above)
Comment (marrip) [reply workflow/rules/bwa-mem.smk:24]: Do we want to include benchmarking in each tool and make a statistik for it? It might be nice :)

File: workflow/rules/bwa-mem.smk (hunk 1 above)
Comment (marrip) [reply workflow/rules/bwa-mem.smk:12]: I will take a look at the wrapper but maybe that is just the way it should be then. What about the unit wildcard? Is it possible to use it with the wrapper? Btw, how should we go about creating own wrappers?
I guess `snakefmt` should still be applied, right?

Comment (Smeds) [reply workflow/rules/bwa-mem.smk:12]: One  could make change suggestion to the snakemake-wrapper repo, if we want to use fwd and rev as input

File: workflow/rules/bwa-mem.smk (hunk 2 above)
Comment (Smeds) [reply workflow/rules/bwa-mem.smk:24]: I think we should try to add benchmark settings, if one don't want to use it one don't have to activate it during your analysis.

Comment (Smeds) [reply workflow/rules/bwa-mem.smk:24]: maybe set it as 
`benchmark:
    benchmark/alignment/bwa-mem/{sample}.txt
`

Comment (marrip) [reply workflow/rules/bwa-mem.smk:24]: Are these `tsv`-files? Maybe we could be explicit about that 😉

File: workflow/rules/bwa-mem.smk (hunk 3 above)
Comment (Smeds) [reply workflow/rules/bwa-mem.smk:18]: If we want to force snakemake to check for it, we could add as an input.

File: workflow/rules/bwa-mem.smk (hunk 5 above)
Comment (Smeds) [reply workflow/rules/bwa-mem.smk:34]: Sounds good!

File: workflow/rules/bwa-mem.smk (hunk 4 above)
Comment (Smeds) [reply workflow/rules/bwa-mem.smk:38]: `
output:
  {module}/{software}/{file}.bai
log:
   logs/{module}/{software}/{file}_index.txt
`

File: workflow/rules/bwa-mem.smk (hunk 2 above)
Comment (jonca79) [reply workflow/rules/bwa-mem.smk:24]: It is a .tsv file!
There could be other tsv-files generated by the rules. I suggest that we add .benchmark to the name to make i clearer:
benchmark: "alignment/bwa_mem/{sample}_{unit}.benchmark.tsv"

Comment (marrip) [reply workflow/rules/bwa-mem.smk:24]: good point!

File: workflow/rules/bwa_mem.smk [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    output:
+        bam="alignment/bwa_mem/{sample}_{unit}.bam",
+    params:
+        index=config["references"]["genome"],
+        extra=config["bwa_mem"]["extra"],
+        sort=config["bwa_mem"].get("sort", "samtools"),
+        sort_order=config["bwa_mem"].get("sort_order", "coordinate"),
+        sort_extra="-@ " + str(config["bwa_mem"]["threads"]),
+    log:
+            log="aligment/bwa_mem/{sample}_{unit}.bam.log",
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:23]: I think log does not need a field name `log=`, we probably do not want different log files for the same rule. And the indentation is one tab too many.

File: workflow/rules/bwa_mem.smk [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        config["tools"].get("bwa_mem", config["tools"].get("default", ""))
+    message:
+        "{rule}: Align {wildcards.sample}_{wildcards.unit} with bwa and sort"
+    wrapper:
+        "0.70.0/bio/bwa/mem"
+
+
+rule samtools_index:
+    input:
+        "{module}/{software}/{file}.bam",
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:37]: If we do it like that we could probably skip the wildcards `module` and `software`, right? It could be only `{file}.bam`

File: workflow/rules/bwa_mem.smk [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    output:
+        "{module}/{software}/{file}.bam.bai",
+    log:
+        "{module}/{software}/{file}.bam.bai.log",
+    benchmark:
+        "{module}/{software}/{file}.bam.bai.benchmark.tsv"
+    container:
+        config["tools"]["common"]
+    message:
+        "{rule}: Index {wildcards.module}/{wildcards.software}/{wildcards.file} bam file"
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:47]: Also something that we should agree upon, do we want to indicate the whole path or just a brief description with sample name, unit or something like that. For this rule in particular it should be the whole path, I guess, as there is no wildcard for sample/unit. But maybe in general it is better to have the whole path there anyways?

File: workflow/rules/bwa_mem.smk [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    shell:
+        "samtools index "
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:49]: How do we want to handle multi-line commands? Should it be quotes in every line or should we do triple quotes for start and end and connect the lines with a forward-slash (`\`). It might not be super important but maybe we can discuss.

File: workflow/rules/common.smk [hunk 11]
```diff
@@ -28,9 +28,7 @@ validate(samples, schema="../schemas/samples.schema.yaml")
 
 ### Read and validate units file
 
-units = pd.read_table(config["units"], dtype=str).set_index(
-    ["sample", "unit", "run", "lane"], drop=False
-)
+units = pd.read_table(config["units"], dtype=str).set_index(["sample", "unit", "run", "lane"], drop=False)
```
Comment (marrip) [inline workflow/rules/common.smk:31]: Did you run a snakefmt on it?

File: workflow/rules/bwa_mem.smk [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+rule bwa_mem:
+    input:
+        reads=["pre-alignment/merge/{sample}_{unit}_R1.fq.gz", "Pre-alignment/merge/{sample}_{unit}_R2.fq.gz"],
+        index=config["references"]["genome"],
+    output:
+        bam="alignment/bwa_mem/{sample}_{unit}.bam",
+    params:
+        index=config["references"]["genome"],
+        extra=config["bwa_mem"]["extra"],
+        sort=config["bwa_mem"].get("sort", "samtools"),
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:19]: I was wondering how `.get` works, does it use the default if the field is `""` or if it does not exist?

File: workflow/rules/bwa_mem.smk [hunk 13]
```diff
@@ -0,0 +1,51 @@
+# vim: syntax=python tabstop=4 expandtab
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:1]: I would remove the `vim`-header as people might use different editors.

File: workflow/rules/common.smk (hunk 11 above)
Comment (jonca79) [reply workflow/rules/common.smk:31]: Yes, it said to change it to this.
snakefmt --compact-diff -l 130

Comment (marrip) [reply workflow/rules/common.smk:31]: ok, maybe I have a different version. Which version do you use?

File: workflow/rules/bwa_mem.smk (hunk 12 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk:19]: If first argument does not exist in the hash it returns the second argument.

File: workflow/rules/common.smk (hunk 11 above)
Comment (jonca79) [reply workflow/rules/common.smk:31]: 0.4.0

File: workflow/rules/bwa_mem.smk (hunk 12 above)
Comment (marrip) [reply workflow/rules/bwa_mem.smk:19]: hm, so empty string would be a valuable argument for get?

File: workflow/rules/common.smk (hunk 11 above)
Comment (marrip) [reply workflow/rules/common.smk:31]: Ah, ok, I see they have 0.4.3 as latest. But could be it doesn't change anything.

File: workflow/rules/bwa_mem.smk (hunk 7 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk:23]: done

File: workflow/rules/bwa_mem.smk (hunk 12 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk:19]: A hash in python can't have  "" as a key so only the default value can be ""

Comment (marrip) [reply workflow/rules/bwa_mem.smk:19]: ok 👍

File: .tests/integration/config.yaml [hunk 14]
```diff
@@ -5,9 +5,9 @@ tools:
...
-   genome: "/data/ref_genomes/hg19/bwa/BWA_0.7.10_refseq/hg19.with.mt.fasta"
+references:
+  genome: "/data/ref_genomes/hg19/bwa/BWA_0.7.10_refseq/hg19.with.mt.fasta"
 
- bwa-mem:
-   threads: 2
-   extra: r"-R '@RG\tID:{sample}\tSM:{sample}\tPL:illumina\tPU:{sample} -v 1 '"
+bwa_mem:
+  threads: 2
+  extra: r"-R '@RG\tID:{sample}\tSM:{sample}\tPL:illumina\tPU:{sample} -v 1 '"
```
Comment (marrip) [inline .tests/integration/config.yaml:13]: Just thinking that maybe we should have `params: extra:` to follow the snakemake rule fields?

File: .tests/integration/config.yaml [hunk 15]
```diff
@@ -5,9 +5,9 @@ tools:
   common: "docker://marrip/common:1.1.0"
   default: "docker://gmsuppsala/somatic:develop"
 
- references:
-   genome: "/data/ref_genomes/hg19/bwa/BWA_0.7.10_refseq/hg19.with.mt.fasta"
+references:
+  genome: "/data/ref_genomes/hg19/bwa/BWA_0.7.10_refseq/hg19.with.mt.fasta"
```
Comment (marrip) [inline .tests/integration/config.yaml:9]: I was wondering where this file is located because it is not part of the repo, right?

Comment (jonca79) [reply .tests/integration/config.yaml:9]: Its on Marvin at the location specified.

File: .tests/integration/config.yaml (hunk 14 above)
Comment (jonca79) [reply .tests/integration/config.yaml:13]: I agree.
Can you have that on one row like this:
bwa_mem:
  params: extra: r"-R '@RG\tID:{sample}\tSM:{sample}\tPL:illumina\tPU:{sample} -v 1 '"

File: .tests/integration/config.yaml (hunk 15 above)
Comment (marrip) [reply .tests/integration/config.yaml:9]: ah, ok, is it to run the jenkins integration then?

File: .tests/integration/config.yaml (hunk 14 above)
Comment (marrip) [reply .tests/integration/config.yaml:13]: I would probably do:

```
params:
  extra: ...
```

In case there are several params we want to set through the `config.yaml`.

File: workflow/rules/bwa_mem.smk (hunk 13 above)
Comment (Smeds) [reply workflow/rules/bwa_mem.smk]: Would it be a problem to leave it? I use vim a lot

File: workflow/rules/bwa_mem.smk [hunk 16]
```diff
@@ -8,7 +8,7 @@ __license__ = "GPL-3"
 
 rule bwa_mem:
     input:
-        reads=["pre-alignment/merge/{sample}_{unit}_R1.fq.gz", "Pre-alignment/merge/{sample}_{unit}_R2.fq.gz"],
+        reads=["pre-alignment/merge/{sample}_{unit}_R1.fq.gz", "pre-alignment/merge/{sample}_{unit}_R2.fq.gz"],
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk:11]: I assume we leave the name `pre-alignment` then? Or could we change it to `prealignment`?

Comment (Smeds) [reply workflow/rules/bwa_mem.smk:11]: The pre-alignment module will use "prealignment".

File: .tests/integration/config.yaml [hunk 17]
```diff
@@ -3,3 +3,12 @@ units: "units.tsv"
 
 tools:
   common: "docker://marrip/common:1.1.0"
+
+reference:
+  fasta: "reference/WASH7P.fna"
+
+bwa_mem:
+  container: "docker://gmsuppsala/somatic:develop"
+  threads: 2
```
Comment (Smeds) [inline .tests/integration/config.yaml:12]: did we decide that threads would be handled by a profile?

File: workflow/rules/common.smk [hunk 18]
```diff
@@ -50,6 +50,6 @@ def get_sample_fastq(wildcards):
 
 def compile_output_list(wildcards):
     output_list = []
-    output_list += expand("alignment/bwa_mem/{sample}_{unit}.bam", sample=samples.index)
-    output_list += expand("alignment/bwa_mem/{sample}_{unit}.bam.bai", sample=samples.index)
+    output_list += expand("alignment/bwa_mem/{sample}_{unit}.bam", sample=samples.index, unit=units.unit)
```
Comment (marrip) [inline workflow/rules/common.smk:53]: I am not sure if this works. I think what `expand` does, it takes both lists of items and puts out all possible combinations which is not exactly what the target should be. for each `sample` only the available `unit` should be used.

File: workflow/rules/bwa_mem.smk (hunk 16 above)
Comment (marrip) [reply workflow/rules/bwa_mem.smk:11]: 👍

Comment (jonca79) [reply workflow/rules/bwa_mem.smk:11]: changed in new version

File: workflow/Snakefile [hunk 19]
```diff
@@ -1,26 +1,17 @@
-# vim: syntax=python tabstop=4 expandtab
 # coding: utf-8
 
 __author__ = "Patrik Smeds"
```
Comment (Smeds) [inline workflow/Snakefile:3]:
```suggestion
__author__ = "Jonas Almlöf"
```

File: workflow/Snakefile [hunk 20]
```diff
@@ -1,26 +1,17 @@
...
 __copyright__ = "Copyright 2021, Patrik Smeds"
```
Comment (Smeds) [inline workflow/Snakefile:4]:
```suggestion
__copyright__ = "Copyright 2021,  Jonas Almlöf"
```

File: workflow/rules/bwa_mem.smk (hunk 13 above)
Comment (marrip) [reply workflow/rules/bwa_mem.smk]: Me too, but with the snakemake syntaxing I do not need any headers. In my opinion we should separate concerns and remove anything that is not part of the code base. But if it impairs your work heavily we can keep it. 😉

File: workflow/Snakefile [hunk 21]
```diff
@@ -1,26 +1,17 @@
...
 __email__ = "patrik.smeds@scilifelab.uu.se"
```
Comment (Smeds) [inline workflow/Snakefile:5]:
```suggestion
__email__ = "jonas.almlof@scilifelab.uu.se"
```

File: workflow/rules/bwa_mem.smk [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+__author__ = "Jonas Almlöf"
+__copyright__ = "Copyright 2021, Jonas Almlöf"
+__email__ = "jonas.almlof@scilifelab.uu.se"
+__license__ = "GPL-3"
+
+
+rule bwa_mem:
+    input:
+        reads=["pre-alignment/merge/{sample}_{unit}_R1.fq.gz", "pre-alignment/merge/{sample}_{unit}_R2.fq.gz"],
```
Comment (Smeds) [inline workflow/rules/bwa_mem.smk:11]:
```suggestion
        reads=["prealignment/merged/{sample}_{unit}_fastq1.fq.gz", "prealignment/merged/{sample}_{unit}_fastq2.fq.gz"],
```

Path that prealignment module uses and also changed fq to fastq since we have that in the unit file.

**Review by Smeds — CHANGES_REQUESTED** [review]

File: .tests/integration/config.yaml (hunk 17 above)
Comment (marrip) [reply .tests/integration/config.yaml:12]: I think we need to discuss because there are several ways to handle it. Unfortunately, I was mistaken and the rules do not have direct access to the profile through `snakemake`-magic. I suggest we look at the possibilities, maybe we have some other ideas as well, and then discuss what we think is the best option. I think this post could help us figure something out: https://www.sichong.site/2020/02/25/snakemake-and-slurm-how-to-manage-workflow-with-resource-constraint-on-hpc/

File: .tests/integration/units.tsv [hunk 23]
```diff
@@ -1,2 +1,2 @@
 sample	unit	run	lane	fastq1	fastq2	adapter
-NA12878	N	1	1	input/R1.fq.gz	input/R2.fq.gz	ACGT
+NA12878	N	1	1	prealigned/merged/fastq1.fq.gz	prealigned/merged/fastq2.fq.gz	ACGT
```
Comment (marrip) [inline .tests/integration/units.tsv]: Do you use `prealigned/merged/fastq1.fq.gz` and `prealigned/merged/fastq1.fq.gz` somewhere or why did you change it?

File: workflow/rules/bwa_mem.smk [hunk 24]
```diff
@@ -19,10 +19,17 @@ rule bwa_mem:
         "aligment/bwa_mem/{sample}_{unit}.bam.log",
     benchmark:
         "alignment/bwa_mem/{sample}_{unit}.bam.benchmark.tsv"
+    benchmark:
+        repeat(
+            "aligment/bwa_mem/{sample}_{unit}.output.bam.benchmark.tsv",
+            config.get("rule_name", {}).get("benchmark_repeats", 1),
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk]: Shouldn't `rule_name` be `bwa_mem`?

File: workflow/rules/bwa_mem.smk [hunk 25]
```diff
@@ -19,10 +19,17 @@ rule bwa_mem:
...
+        )
     threads: config["bwa_mem"]["threads"]
     container:
-        config["bwa_mem"]["container"]
+        config.get("bwa_mem", "default_container").get("container", "default_container")
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk]: I assume `default_container` is not a valid container path, right? It needs to be linked to the appropriate field in `config`.

File: workflow/rules/bwa_mem.smk [hunk 26]
```diff
@@ -19,10 +19,17 @@ rule bwa_mem:
...
+    conda:
+        "../envs/rule_name.yaml"
```
Comment (marrip) [inline workflow/rules/bwa_mem.smk]: Again, it should be `bwa_mem`, right?

File: workflow/rules/common.smk [hunk 27]
```diff
@@ -45,16 +45,25 @@ def get_sample_fastq(wildcards):
...
-        if unit.index not in sample_dict:
-            sample_dict[unit.index] = {}
-        if unit not in sample_dict[unit.index]:
-            sample_dict[unit.index][unit] = ""
-    for sample in sample_dict:
-        for unit in sample_dict[sample]:
-            output_list.append("alignment/bwa_mem/%s_%s.bam" % sample, unit)
-            output_list.append("alignment/bwa_mem/%s_%s.bam.bai" % sample, unit)
-    return output_list
+def get_unit_types(units: pd.DataFrame, sample: str) -> set[str]:
```
Comment (marrip) [inline workflow/rules/common.smk]: Can this be imported from the `tools` package?

File: workflow/rules/common.smk [hunk 28]
```diff
@@ -3,12 +3,17 @@ __copyright__ = "Copyright 2021, Jonas Almlöf"
 __email__ = "jonas.almlof@scilifelab.uu.se"
 __license__ = "GPL-3"
 
-import pandas as pd
+
+import pandas as pandas
```
Comment (marrip) [inline workflow/rules/common.smk]: same as in `prealignment`, `as pandas` is probably not required.

File: workflow/rules/common.smk [hunk 29]
```diff
@@ -32,38 +37,13 @@ validate(units, schema="../schemas/units.schema.yaml")
...
-    fastqs = units.loc[(wildcards.sample, wildcards.unit, wildcards.run, wildcards.lane), ["fastq1", "fastq2"]].dropna()
-    return {"fwd": fastqs.fastq1, "rev": fastqs.fastq2}
-
-
-def get_sample_fastq(wildcards):
-    fastqs = units.loc[(wildcards.sample, wildcards.unit), ["fastq1", "fastq2"]].dropna()
-    return {"fwd": fastqs["fastq1"].tolist(), "rev": fastqs["fastq2"].tolist()}
-
-
-def get_unit_types(units: pd.DataFrame, sample: str) -> set[str]:
```
Comment (marrip) [inline workflow/rules/common.smk]: You changed it here 😄

File: workflow/schemas/config.schema.yaml [hunk 30]
```diff
@@ -6,14 +6,38 @@ properties:
...
+  reference:
+    fasta: string
+    required:
+      - fasta
+  bwa_mem:
     type: object
     properties:
-      common:
+      benchmark_repeats:
+        type: string
```
Comment (marrip) [inline workflow/schemas/config.schema.yaml]: Does not not work with `int`?

File: workflow/schemas/config.schema.yaml [hunk 31]
```diff
@@ -16,25 +16,32 @@ properties:
     type: object
     properties:
       benchmark_repeats:
-        type: string
+        type: integer
```
Comment (marrip) [inline workflow/schemas/config.schema.yaml:23]: Haha, and again, you changed it already 👍

File: workflow/rules/bwa_mem.smk (hunk 25 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk]: Will be changed to:
config.get("bwa_mem", {}).get("container", "default_container")

File: workflow/rules/bwa_mem.smk (hunk 26 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk]: It actually workes with:
"../envs/{rule}.yaml"

File: workflow/rules/common.smk (hunk 28 above)
Comment (jonca79) [reply workflow/rules/common.smk]: done

File: workflow/rules/bwa_mem.smk (hunk 24 above)
Comment (jonca79) [reply workflow/rules/bwa_mem.smk]: done

File: .tests/integration/units.tsv (hunk 23 above)
Comment (jonca79) [reply .tests/integration/units.tsv]: Yes I use prealigned/merged/fastq1.fq.gz as this is how the output from the prealign module looks like.

Comment (marrip) [reply .tests/integration/units.tsv]: True, but you are not using these information from the `units.tsv`, right? I would not use these paths here because it creates confusion.

Comment (jonca79) [reply .tests/integration/units.tsv]: Changed back.

File: .tests/integration/units.tsv [hunk 32]
```diff
@@ -1,2 +1,2 @@
 sample	type	run	lane	fastq1	fastq2	adapter
-NA12878	N	1	1	prealignment/merged/NA12878_N_fastq1.fq.gz	prealignment/merged/NA12878_N_fastq2.fq.gz	ACGT
+NA12878	N	1	1	input/R1.fq.gz	input/R2.fq.gz	ACGT
```
Comment (marrip) [inline .tests/integration/units.tsv]: 👍

**Review by marrip — APPROVED** [review]
